### PR TITLE
Fix Java core-models build script

### PR DESCRIPTION
**Pull request by smowton**
This fixes two problems with the core-models build script:

1. Use of add_library with target_source_files listing the same source
   led to including the *source* files along with every build target, hence
   multiple builds of the same file

2. Using add_custom_command to create a pre-build hook instead of a normal
   file-producing command meant java_class_loader.cpp and its dependencies were
   rebuilt on every build.

### CODE REVIEW

**Review by LAJW — APPROVED** [review]
Looks plausible.

**Comment by smowton** [comment]
@cesaro wanted to review?

**Comment by smowton** [comment]
@chrisr-diffblue oddly enough you're the only code owner here, so please review

**Comment by cesaro** [comment]
@smowton Yes, will do tomorrow

File: src/java_bytecode/library/CMakeLists.txt
Comment (chrisr-diffblue) [inline src/java_bytecode/library/CMakeLists.txt]: Should JAVA_CORE_MODELS_INC also use CMAKE_CURRENT_BINARY_DIR here as well? If I'm reading the Cmake correctly, then JAVA_CORE_MODELS_INC is a generated file, so it'd be nicer if that was generated into the build directory rather than the source directory - but maybe that just opens a bigger can of worms than this PR is fixing?

**Review by chrisr-diffblue — APPROVED** [review]
Looks reasonable from my understanding, with one possible improvement which might be nice to fix while you are working on it - but I won't block the rest of the change on that.

File: src/java_bytecode/library/CMakeLists.txt
Comment (smowton) [reply src/java_bytecode/library/CMakeLists.txt]: Yes, and yes. The ideal approach would generate it somewhere, then add that path to the consumer library's include search path. However that's another half-day's cmake studies that I don't have the appetite for just now, so let's address that separately ;)

Comment (chrisr-diffblue) [reply src/java_bytecode/library/CMakeLists.txt]: Understood, and I don't blame you :-)

**Comment by cesaro** [comment]
@smowton As we just spoke, please squash https://github.com/cesaro/cbmc/commit/1a4151aad38a6fc288c0c4281f9615a05cee0570 into your only commit in this PR.

That fixes:
- Generation of the `.inc` file is in the cmake build directory instead of the source directory
- Adds a missing dependency: the `.inc` file also depends on the converter (!)
- Adds documentation explaining how on earth cmake works in these two files ...
- Removes unnecessary variable definitions in the java_bytecode CMakeFile

**Comment by smowton** [comment]
@cesaro done just now, will merge if CI passes

**Pull request merged**
